Sources/SwiftSyntaxMacros/Syntax+LexicalContext.swift

```diff
@@ -0,0 +1,91 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftSyntax
+import SwiftSyntaxBuilder
+
+extension SyntaxProtocol {
+  /// If this syntax node acts as a lexical context from the perspective
+  /// of a macro, return a new syntax node based on this node that strips all
+  /// information that isn't supposed to be exposed as a lexical context, such
+  /// as function bodies or the members of types/extensions.
+  ///
+  /// Returns `nil` for any syntax node that isn't a lexical context.
+  public func asMacroLexicalContext() -> Syntax? {
```

Comment:
Does this have to be public or could we make it internal or SPI?

Reply:
Hmm. We could make this SPI for now, and expose it later if we find a need for it beyond the compiler.

Reply:
Per above comment, we want this as API